### ISSUE TITLE
Services are shown as child devices

### ISSUE THREAD

**Issue by mstoffel-sag** [thin-edge/thin-edge.io#2584]
Services are shown as child devices

![image](https://github.com/thin-edge/thin-edge.io/assets/29426188/f1abe63f-ad2e-4620-befa-b9890ce30eda)

It's weird though that the bridge is show correctly as a service:

![image](https://github.com/thin-edge/thin-edge.io/assets/29426188/ea1b33d1-4f9d-4fca-8429-396ed50bbff7)

**To Reproduce**
I used the tedge-rugpi-image repo to create a pi3 image, lashed it to an sd card and followed the instruction to register the device. 

**Expected behavior**
After registration there are child device that should be displayed as services.

**Environment (please complete the following information):**
 - OS rugpi image release [20231221.2255](https://github.com/thin-edge/tedge-rugpi-image/releases/tag/20231221.2255)
 - pi3b
 - arm64
 - thin-edge.io 
 
![image](https://github.com/thin-edge/thin-edge.io/assets/29426188/3e41cbc1-4510-44d4-8f8c-68257847673d)

**Comment by reubenmiller** [comment]
@mstoffel-sag Can you please add the output of the MQTT messages using the following command (wait 5 seconds, then kill the command, then post the console output)

```sh
tedge mqtt sub 'te/+/+/+/+'
```

We suspect that there is something incorrect with the order of messages which is causing the services to be registered a child devices.

**Comment by mstoffel-sag** [comment]
Here's the output:

root@rpi3-b827eb596700 ~ # tedge mqtt sub 'te/+/+/+/+'
INFO: Connected
[te/device/main/service/tedge-mapper-collectd] {"@parent":"device/main//","@type":"service","type":"service"}
[te/device/main/service/tedge-agent] {"@parent":"device/main//","@type":"service","type":"service"}
[te/device/main/service/tedge-mapper-c8y] {"@parent":"device/main//","@type":"service","type":"service"}
[te/device/main/service/mosquitto-c8y-bridge] {"@id":"rpi3-b827eb596700:device:main:service:mosquitto-c8y-bridge","@parent":"device/main//","@type":"service","name":"mosquitto-c8y-bridge","type":"service"}

**Comment by reubenmiller** [comment]
Thanks @mstoffel-sag. The team has been able to reproduce the issue and analyzing the results.

**Comment by Bravo555** [comment]
The issue is triggered by SmartREST message [`104`](https://cumulocity.com/guides/reference/smartrest-two/#service-status-update-104). Despite being described as _"Set a status for given software service."_ , if `104` is sent before registering a service, it actually creates a child device.

**Comment by reubenmiller** [comment]
> The issue is triggered by SmartREST message [`104`](https://cumulocity.com/guides/reference/smartrest-two/#service-status-update-104). Despite being described as _"Set a status for given software service."_ , if `104` is sent before registering a service, it actually creates a child device.

Good find. So we're going to have to make sure we hold off sending 104 messages until the 102 is sent.

**Comment by reubenmiller** [comment]
@Bravo555 @albinsuresh 

After investigating the image created by [tedge-rugpi-image](https://github.com/thin-edge/tedge-rugpi-image/tree/next), it seems that disable the services before `tedge connect c8y` is run seems to fix some of the services being created, however there is an additional side effect now where an additional child device called `operations` was created erroneously, though I created a separate ticket for this: https://github.com/thin-edge/thin-edge.io/issues/2606

So it seems that starting up the services before `tedge connect c8y` is run is definitely causing unexpected issues (that need to be fixed).

**Comment by reubenmiller** [comment]
## Procedure

The following repositories/branches were used to build test images.

|Name|Branch|
|----|------|
|tedge-rugpi-image|[feat-debug-example](https://github.com/thin-edge/tedge-rugpi-image/tree/feat-debug-example)|
|tedge-rugpi-core|[feat-debug-logger](https://github.com/thin-edge/tedge-rugpi-core/tree/feat-debug-logger)|

Then the following procedure was used for each test case:

1. Delete the existing device

  ```sh
  c8y devices delete --id rpi4-d83add90fe56 --cascade
  ```

2. Delete entries from the known_hosts file

    ```sh
    sed -i '/rpi4-d83add90fe56/d' /Users/reubenmiller/.ssh/known_hosts
    ```

3. Create the rugpi image

4. Flash image to SD card

5. SSH onto device and check that services are either started/stopped

    ```
    ssh root@rpi4-d83add90fe56.local
    ```

    **Check service status**

    ```sh
    systemctl status tedge-agent
    systemctl status tedge-mapper-c8y
    systemctl status tedge-mapper-collectd
    systemctl status "tedge-*"
    ```

6. Create a bootstrapping entry (to make it visible in the logs when bootstrapping is occurring)

    ```sh
    tedge mqtt pub 'debug/bootstrapping' 'bootstrapping soon'
    ```

7. Bootstrap device

    ```sh
    c8y tedge bootstrap root@rpi4-d83add90fe56.local
    ```

8. Copy the mqtt log output to your local machine

    ```sh
    scp root@rpi4-d83add90fe56.local:/data/mqtt.log ./mqtt-autostart-disabled.log
    ```

## Test case 1: Auto start disabled

[mqtt-autostart-disabled.log](https://github.com/thin-edge/thin-edge.io/files/14009017/mqtt-autostart-disabled.log)

### Before bootstrapping

```log
# tedge mqtt sub '#'
INFO: Connected
[te/device/main///twin/c8y_Hardware] {"sysClockSync":"no","model":"Raspberry Pi 4 Model B Rev 1.5","revision":"-d03115","serialNumber":"1000000044d70329"}
[te/device/main///twin/firmware] {"name":"tedge_rugpi_tryboot","version":"20240122.1045","updated":"Mon 11 Dec 04:55:24 GMT 2023"}
[te/device/main///twin/device_OS] {"displayName":"Debian GNU/Linux 12 (bookworm)","version":"12","hostname":"rpi4-d83add90fe56","arch":"aarch64","family":"GNU/Linux","kernel":"#1 SMP PREEMPT Debian 1:6.1.63-1+rpt1 (2023-11-24)","sysClockSync":"no"}
```

**Note** `tedge-inventory` service is still configured for autostart


## Test case 2: Auto start enabled

[mqtt-autostart-enabled.log](https://github.com/thin-edge/thin-edge.io/files/14009018/mqtt-autostart-enabled.log)

### Before bootstrapping

```log
# tedge mqtt sub '#'
INFO: Connected
[te/device/main///cmd/config_snapshot] {"types":["tedge-configuration-plugin","tedge-log-plugin","tedge-mosquitto.conf","tedge.toml"]}
[te/device/main///cmd/config_update] {"types":["tedge-configuration-plugin","tedge-log-plugin","tedge-mosquitto.conf","tedge.toml"]}
[te/device/main///cmd/log_upload] {"types":["c8y_Command","firmware_update","software-management"]}
[te/device/main///cmd/firmware_update] {}
[te/device/main///cmd/restart] {}
[te/device/main///cmd/software_list] {}
[te/device/main///cmd/software_update] {}
[te/device/main///twin/device_OS] {"displayName":"Debian GNU/Linux 12 (bookworm)","version":"12","hostname":"rpi4-d83add90fe56","arch":"aarch64","family":"GNU/Linux","kernel":"#1 SMP PREEMPT Debian 1:6.1.63-1+rpt1 (2023-11-24)","sysClockSync":"no"}
[te/device/main///twin/firmware] {"name":"tedge_rugpi_tryboot","version":"20240122.1021","updated":"Mon 11 Dec 04:55:20 GMT 2023"}
[te/device/main///twin/c8y_Hardware] {"sysClockSync":"no","model":"Raspberry Pi 4 Model B Rev 1.5","revision":"-d03115","serialNumber":"1000000044d70329"}
[te/device/main/service/tedge-agent] {"@parent":"device/main//","@type":"service","type":"service"}
[te/device/main/service/tedge-agent/status/health] {"pid":772,"status":"up","time":1702270504.030202}
[te/device/main/service/tedge-mapper-collectd] {"@parent":"device/main//","@type":"service","type":"service"}
[te/device/main/service/tedge-mapper-collectd/status/health] {"pid":794,"status":"up","time":1702270504.0413673}
```

### After bootstrapping

```
# tedge mqtt sub '#'
INFO: Connected
[te/device/main///cmd/config_snapshot] {"types":["tedge-configuration-plugin","tedge-log-plugin","tedge-mosquitto.conf","tedge.toml"]}
[te/device/main///cmd/config_update] {"types":["tedge-configuration-plugin","tedge-log-plugin","tedge-mosquitto.conf","tedge.toml"]}
[te/device/main///cmd/log_upload] {"types":["c8y_Command","firmware_update","software-management"]}
[te/device/main///cmd/firmware_update] {}
[te/device/main///cmd/restart] {}
[te/device/main///cmd/software_list] {}
[te/device/main///cmd/software_update] {}
[te/device/main///twin/device_OS] {"displayName":"Debian GNU/Linux 12 (bookworm)","version":"12","hostname":"rpi4-d83add90fe56","arch":"aarch64","family":"GNU/Linux","kernel":"#1 SMP PREEMPT Debian 1:6.1.63-1+rpt1 (2023-11-24)","sysClockSync":"no"}
[te/device/main///twin/firmware] {"name":"tedge_rugpi_tryboot","version":"20240122.1021","updated":"Mon 11 Dec 04:55:20 GMT 2023"}
[te/device/main///twin/c8y_Hardware] {"sysClockSync":"no","model":"Raspberry Pi 4 Model B Rev 1.5","revision":"-d03115","serialNumber":"1000000044d70329"}
[te/device/main///twin/c8y_Agent] {"name":"thin-edge.io","url":"https://thin-edge.io","version":"1.0.0-rc.2~116+g02d20ea"}
[te/device/main/service/tedge-agent] {"@parent":"device/main//","@type":"service","type":"service"}
[te/device/main/service/tedge-agent/status/health] {"pid":772,"status":"up","time":1705916358.6633723}
[te/device/main/service/tedge-mapper-collectd] {"@parent":"device/main//","@type":"service","type":"service"}
[te/device/main/service/tedge-mapper-collectd/status/health] {"pid":794,"status":"up","time":1705916358.663361}
[te/device/main/service/tedge-mapper-c8y] {"@parent":"device/main//","@type":"service","type":"service"}
[te/device/main/service/tedge-mapper-c8y/status/health] {"pid":2349,"status":"up","time":1705916358.6610012}
[te/device/main/service/mosquitto-c8y-bridge] {"@id":"rpi4-d83add90fe56:device:main:service:mosquitto-c8y-bridge","@parent":"device/main//","@type":"service","name":"mosquitto-c8y-bridge","type":"service"}
[te/device/main/service/mosquitto-c8y-bridge/status/health] 1
```


**file: entity_store.jsonl**

```json
{"version":"1.0"}
{"topic":{"name":"te/device/main///twin/c8y_Agent"},"payload":"{\"name\":\"thin-edge.io\",\"url\":\"https://thin-edge.io\",\"version\":\"1.0.0-rc.2~116+g02d20ea\"}","qos":1,"retain":true}
{"topic":{"name":"te/device/main///twin/c8y_Hardware"},"payload":"{\"model\":\"Raspberry Pi 4 Model B Rev 1.5\",\"revision\":\"-d03115\",\"serialNumber\":\"1000000044d70329\",\"sysClockSync\":\"no\"}","qos":1,"retain":true}
{"topic":{"name":"te/device/main///twin/c8y_Firmware"},"payload":"{\"name\":\"tedge_rugpi_tryboot\",\"updated\":\"Mon 11 Dec 04:55:24 GMT 2023\",\"version\":\"20240122.1045\"}","qos":1,"retain":true}
{"topic":{"name":"te/device/main///twin/device_OS"},"payload":"{\"arch\":\"aarch64\",\"displayName\":\"Debian GNU/Linux 12 (bookworm)\",\"family\":\"GNU/Linux\",\"hostname\":\"rpi4-d83add90fe56\",\"kernel\":\"#1 SMP PREEMPT Debian 1:6.1.63-1+rpt1 (2023-11-24)\",\"sysClockSync\":\"no\",\"version\":\"12\"}","qos":1,"retain":true}
{"topic":{"name":"te/device/main/service/mosquitto-c8y-bridge"},"payload":"{\"@id\":\"rpi4-d83add90fe56:device:main:service:mosquitto-c8y-bridge\",\"@parent\":\"device/main//\",\"@type\":\"service\",\"name\":\"mosquitto-c8y-bridge\",\"type\":\"service\"}","qos":1,"retain":true}
{"topic":{"name":"te/device/main/service/tedge-mapper-c8y"},"payload":"{\"@parent\":\"device/main//\",\"@type\":\"service\",\"type\":\"service\"}","qos":1,"retain":true}
{"topic":{"name":"te/device/main/service/tedge-agent"},"payload":"{\"@parent\":\"device/main//\",\"@type\":\"service\",\"type\":\"service\"}","qos":1,"retain":true}
```

**Comment by mstoffel-sag** [comment]
Just wondering, is that maybe also a "wrong" behaviour of core?

**Comment by reubenmiller** [comment]
> Just wondering, is that maybe also a "wrong" behaviour of core?

I wouldn't say so, but yes the core's behaviour makes this more difficult than it should be

**Comment by didier-wenzek** [comment]
I observed the same issue on a device where thin-edge was running for a while without being connected to any c8y endpoint. When finally connected to a new tenant (as I did after the staging tenant has been wiped) or to the same tenant but with a new device id (as I tested), then the service health status messages are forwarded by the mapper to c8y (104) without any service registration (102).

This is not an expected use-case, but this highlights that the registration mechanism is really fragile.

Would it make sense to send a smartrest 102 message before all 104 messages?

**Comment by Bravo555** [comment]
The issue should be fixed in #2620 by changing service status smartrest message from 104 to 102, with the 102 actually creating a service if it didn't exist yet.
A new test suite `bootstrap.robot` was created with 2 tests checking if no unexpected child devices are created. The first case relates to this ticket, and the second one to #2606, although the second case doesn't work correctly because fs events are not properly detected by inotify inside the test device container.

**Comment by gligorisaev** [comment]
QA has thoroughly checked the bug and here are the results:

- [x] Test for ticket exists in the test suite.
- tests/RobotFramework/tests/cumulocity/bootstrap.robot
- [x] QA has tested the bug and it's not reproducable anymore
